dataset/data_array.cpp

```diff
@@ -90,9 +90,14 @@ DataArray DataArray::slice(const Slice &s) const {
           m_attrs->slice(s).merge_from(out_attrs), m_name};
 }
 
-DataArray &DataArray::setSlice(const Slice &s, const DataArray &array) {
+void DataArray::validateSlice(const Slice &s, const DataArray &array) const {
   expect::coordsAreSuperset(slice(s), array);
-  // TODO Need dry-run mechanism for mask handling?
+  data().validateSlice(s, array.data());
+  masks().validateSlice(s, array.masks());
+}
+
+DataArray &DataArray::setSlice(const Slice &s, const DataArray &array) {
+  validateSlice(s, array);
   masks().setSlice(s, array.masks());
```

Comment:
On our short call this morning we had concluded that we are avoiding double-checking things such as coords, but here we actually check masks a second time? Maybe we should also validate coords and data directly here, then attempt `masks().setSlice()` which would call the same as the last line of `DataArray::validateSlice` above?

Reply:
I'm not clear on what you are suggesting here. Maybe best discussed on a call.